Mathlib/AlgebraicGeometry/Nilpotents.lean

```diff
@@ -0,0 +1,212 @@
+  exact X.toΓSpecFun_bijective_of_isAffine.surjective
+
+/-- If `X` is an affine scheme, every closed set of `X` is the zero locus
+of a set of global sections. -/
+lemma eq_zeroLocus_of_isClosed (X : Scheme) [IsAffine X] (s : Set X) (hs : IsClosed s) :
+    ∃ I : Ideal (Γ(X, ⊤)), s = X.toRingedSpace.zeroLocus (I : Set Γ(X, ⊤)) := by
+  let Z : Set (Spec <| Γ(X, ⊤)) := X.toΓSpecFun '' s
+  have hZ : IsClosed Z := X.toΓSpecFun_isClosedMap _ hs
+  obtain ⟨I, (hI : Z = _)⟩ := (PrimeSpectrum.isClosed_iff_zeroLocus_ideal _).mp hZ
+  use I
+  simp only [← LocallyRingedSpace.toΓSpec_preim_zeroLocus_eq, ← hI, Z]
+  rw [Set.preimage_image_eq _ X.toΓSpecFun_bijective_of_isAffine.injective]
+
+end ZeroLocus
+
+section Nilpotents
+
+/-- If a global section is nilpotent on each member of a finite open cover, then `f` is
+nilpotent. -/
+lemma Scheme.isNilpotent_of_isNilpotent_cover {X : Scheme.{u}} (s : Γ(X, ⊤)) (𝒰 : X.OpenCover)
+    [Finite 𝒰.J] (h : ∀ i : 𝒰.J, IsNilpotent (Scheme.Γ.map (𝒰.map i).op s)) : IsNilpotent s := by
+  choose fn hfn using h
+  have : Fintype 𝒰.J := Fintype.ofFinite 𝒰.J
+  /- the maximum of all `fn i` (exists, because `𝒰.J` is finite) -/
+  let N : ℕ := Finset.sup Finset.univ fn
+  have hfnleN (i : 𝒰.J) : fn i ≤ N := Finset.le_sup (Finset.mem_univ i)
+  use N
+  apply zero_of_zero_cover
+  intro i
+  simp only [map_pow]
+  exact pow_eq_zero_of_le (hfnleN i) (hfn i)
+
+/-- If `U` is an affine open of a scheme `X`, a section over `U` is nilpotent if and only if
+the associated basic open is empty.
+See `Scheme.isNilpotent_iff_basicOpen_eq_bot_of_isCompact` for a version only assuming
+`U` to be compact. -/
+lemma IsAffineOpen.isNilpotent_iff_basicOpen_eq_bot {X : Scheme} {U : Opens X} (hU : IsAffineOpen U)
+    (f : Γ(X, U)) : IsNilpotent f ↔ X.basicOpen f = ⊥ := by
+  constructor
+  · intro h
+    apply X.basicOpen_eq_bot_of_isNilpotent
+    exact h
+  · intro h
+    rw [← PrimeSpectrum.basicOpen_eq_bot_iff, ← hU.fromSpec_preimage_basicOpen, h]
+    rfl
+
+/-- If `X` is an affine scheme and `f` is a global section, then `f` is nilpotent if and only
+if the associated basic open is empty. -/
+lemma Scheme.isNilpotent_iff_basicOpen_eq_bot_of_isAffine (X : Scheme) [IsAffine X] (f : Γ(X, ⊤)) :
+    IsNilpotent f ↔ X.basicOpen f = ⊥ :=
+  (isAffineOpen_top X).isNilpotent_iff_basicOpen_eq_bot f
+
+instance Scheme.ιOpens_appLE_isIso {X : Scheme.{u}} (U : Opens X) :
+    IsIso ((Scheme.ιOpens U).appLE U ⊤ (by simp)) := by
+  simp only [restrict_presheaf_obj, ofRestrict_appLE]
+  let e : (Scheme.ιOpens U) ''ᵁ ⊤ ≤ U := by simp
+  have : IsIso (homOfLE e) := by
+    apply homOfLE_isIso_of_eq
+    simp
+  apply Functor.map_isIso
+
+/-- If `U` is a compact open of a scheme and `f` a section over `U` such that the associated
+basic open is empty, then `f` is nilpotent. This is the harder direction of
+`Scheme.isNilpotent_iff_basicOpen_eq_bot_of_isCompact`. -/
+lemma Scheme.isNilpotent_of_basicOpen_eq_bot_of_isCompact {X : Scheme.{u}}
```

Comment:
Is `exists_pow_mul_eq_zero_of_res_basicOpen_eq_zero_of_isCompact` related?

Reply:
Yes, I now conclude this lemma from the one you linked.